course6/string_ariel/str.h

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <iostream>
+#include <cstring>
+
+class Str
+{
+public:
+    Str()
+    {
+        start = mid = end = nullptr;
+    }
+
+    Str(const char* s)
+    {
+        int size = std::strlen(s);
+        start = new char[size];
+        end = start;
+
+        while (*s != '\0')
+        {
+            *end++ = *s++;
+        }
+
+        mid = end;
+    }
+
+    Str(const Str& str)
+    {
+        int size = str.Size();
+        start = new char[size];
+        end = start;
+
+        for (int i = 0; i < size; i++)
+        {
+            *end++ = str[i];
+        }
+
+        mid = end;
+    }
+
+    Str(Str&& str) :start(str.start), mid(str.mid), end(str.end)
+    {
+        str.start = str.mid = str.end = nullptr;
+    }
+
+    ~Str()
+    {
+        Dispose();
+    }
+
+    char operator[](int i) const
+    {
+        return *(start + i);
+    }
+
+    char& operator[](int i)
+    {
+        return *(start + i);
+    }
+
+    Str& operator = (const Str& str)
+    {
+        if (str.Size() > Capacity())
+        {
+            Str temp(str);
+            Swap(*this, temp);
+        }
+        else
+        {
+            mid = start + str.Size();
+            Copy(str);
+        }
+
+        return *this;
+    }
+
+    Str& operator = (const char* s)
+    {
+        int size = std::strlen(s);
+        if (size > Capacity())
+        {
+            Str temp(s);
+            Swap(*this, temp);
+        }
+        else
+        {
+            mid = start + size;
+            Copy(s);
+        }
+
+        return *this;
+    }
+
+    Str operator + (const Str& rh)
+    {
+        return Str(*this, rh);
+    }
+
+    Str& operator += (const Str& rh)
+    {
+        if(rh.Size() + Size() <= Capacity())
+        {
+            for(int i = 0; i < rh.Size(); i++)
+            {
+                *mid++ = rh[i];
+            }
+        }
+        else
+        {
+            Str temp = *this + rh;
+            Swap(*this, temp);
+        }
+
+        return *this;
+    }
+
+    int Size() const
```

Comment:
When implementing container, stick to the standard naming and interface. This way you can write generic code that can work on different containers, including yours.
Specifically, you should write your unit test as a template method, and it should pass for both your string and ```std::string```.

Reply:
I tried to stay away as possible from container and some std stuff that we haven’t seen in the classes